File: data_structures/binary_tree/bin_tree.py


class BinaryTree:

    class Node:
        def __init__(self, data):
            self.data = data
            self.left = self.right = None

    def __init__(self, values=[]):
        self.root = None
        self.root = self.build_tree(self.root, 0, values)
        self.sorted_root = None
        self.sorted_root = self.build_tree_sorted(self.sorted_root, values)

    def build_tree(self, root, itr, values):
        if itr < len(values):
            root = self.Node(values[itr])
            root.left = self.build_tree(root.left, itr * 2 + 1, values)
            root.right = self.build_tree(root.right, itr * 2 + 2, values)
        return root
    
    def insert(self, node, data):
        if node == None:
            return self.Node(data)
        if data <= node.data:
            node.left = self.insert(node.left, data)
        elif data > node.data:
            node.right = self.insert(node.right, data)
        return node

    def build_tree_sorted(self, root, values):
        root = self.insert(root, values[0])
        for i in range(len(values)):
            self.insert(root, values[i])
        return root
    
    def apply_preorder(self, node, fn):
        if node:
            fn(node)
            self.apply_preorder(node.left, fn)
            self.apply_preorder(node.right, fn)

    def apply_inorder(self, node, fn):
        if node:
            self.apply_inorder(node.left, fn)
            fn(node)
            self.apply_inorder(node.right, fn)
    
    def apply_postorder(self, node, fn):
        if node:
            self.apply_postorder(node.left, fn)
            self.apply_postorder(node.right, fn)
            fn(node)
            




def print_node(node):
    print(node.data, end=" ")



btree = BinaryTree([1, 1, 1, 1, 1])

# btree.apply_preorder(btree.root, print_node)
# print()
btree.apply_inorder(btree.sorted_root, print_node)
print()
# btree.apply_postorder(btree.root, print_node)

File: data_structures/binary_tree/binTree.ts
class _Node {
    constructor(
        public data: number,
        public left: _Node = null,
        public right: _Node = null){}
}

class BinaryTree {
    private _root: _Node;
    private _sortedRoot: _Node;
    get root(): _Node {
        return this._root;
    }
    get sortedRoot(): _Node {
        return this._sortedRoot;
    }

    constructor(values: number[] = []){
        this._root = this.build_tree(this._root, values, 0);
        this._sortedRoot = this.build_tree_sorted(this._sortedRoot, values);

    }

    private build_tree(node: _Node, values: number[], itr: number): _Node {
        if (itr < values.length) {
            node = new _Node(values[itr]);
            node.left = this.build_tree(node.left, values, itr * 2 + 1);
            node.right = this.build_tree(node.right, values, itr * 2 + 2);
        }
        return node;
    }

    private insert(node: _Node, value: number): _Node {
        if (node == null) {
            return new _Node(value);
        }
        if (value <= node.data) {
            node.left = this.insert(node.left, value);
        }
        else if (value > node.data){
            node.right = this.insert(node.right, value);
        }
        return node;
    }

    private build_tree_sorted(root: _Node, values: number[]): _Node {
        root = null
        root = this.insert(root, values[0])
        for(let i: number = 1; i < values.length; i++){
            this.insert(root, values[i])
        }
        return root;
    }

    public apply_preorder(node: _Node, fn: (node: _Node) => void) {
        if (node != null)
        {
            fn(node);
            this.apply_preorder(node.left, fn);
            this.apply_preorder(node.right, fn);
        }
    }
    public apply_inorder(node: _Node, fn: (node: _Node) => void) {
        if (node != null)
        {

            this.apply_inorder(node.left, fn);
            fn(node);
            this.apply_inorder(node.right, fn);
        }
    }

    public apply_postorder(node: _Node, fn: (node: _Node) => void) {
        if (node != null)
        {
            this.apply_postorder(node.left, fn);
            this.apply_postorder(node.right, fn);
            fn(node);
        }
    }
}

function print_node_value(node:_Node): void {
    console.log(node.data);
}


var values = [1, 1, 1, 1, 1]

const btree = new BinaryTree(values);

btree.apply_preorder(btree.sortedRoot, print_node_value);
console.log()
btree.apply_inorder(btree.sortedRoot, print_node_value);
console.log()
btree.apply_postorder(btree.sortedRoot, print_node_value);

